test: reconfigure how runtimes are passed in

haircommander opened the pull request:
<!--  Thanks for sending a pull request!

Please be aware that we're following the Kubernetes guidelines of contributing
to this project. This means that we have to use this mandatory template for all
of our pull requests.

Please also make sure you've read and understood our contributing guidelines
(https://github.com/cri-o/cri-o/blob/master/CONTRIBUTING.md) as well as ensuring
that all your commits are signed with `git commit -s`.

Here are some additional tips for you:

- If this is your first time, please read our contributor guidelines:
  https://git.k8s.io/community/contributors/guide#your-first-contribution and
  developer guide
  https://git.k8s.io/community/contributors/devel/development.md#development-guide
- Please label this pull request according to what type of issue you are
  addressing, especially if this is a release targeted pull request. For
  reference on required PR/issue labels, read here:
  https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
- If you want *faster* PR reviews, read how:
  https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
- If the PR is unfinished, see how to mark it:
  https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Uncomment only one `/kind <>` line, hit enter to put that in a new line, and
remove leading whitespace from that line:
-->

 /kind cleanup


#### What this PR does / why we need it:
As well as update documentation describing how to specify an alternative runtime
And fix some inconsistencies in how the runtime is called in restore.bats

Signed-off-by: Peter Hunt <pehunt@redhat.com>
#### Which issue(s) this PR fixes:

<!--
Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->

<!--
Fixes #
or
None
-->

#### Special notes for your reviewer:
replaces https://github.com/cri-o/cri-o/pull/4099
#### Does this PR introduce a user-facing change?

<!--
If no, just write `None` in the release-note block below. If yes, a release note
is required: Enter your extended release note in the block below. If the PR
requires additional action from users switching to the new release, include the
string "action required".

For more information on release notes see:
https://git.k8s.io/community/contributors/guide/release-notes.md
-->

```release-note
none
```

haircommander commented:
@fgiudici  PTAL
the advantage here is we only need to specify 
`CONTAINER_DEFAULT_RUNTIME=crun` and the test takes care of the rest (or also RUNTIME_TYPE if it's vm based)

fgiudici commented on test/README.md:
````diff
@@ -70,11 +70,16 @@ Tests on the host will run with `runc` as the default runtime.
+For example, to use [kata](https://github.com/kata-containers/runtime) with shim v2:
 ```
-make localintegration RUNTIME=cc-runtime
+make CONTAINER_DEFAULT_RUNTIME=kata RUNTIME_TYPE=vm localintegration
````
The kata shim v2 binary has a slightly cumbersome name ;-)
```suggestion
make CONTAINER_DEFAULT_RUNTIME=containerd-shim-kata-v2 RUNTIME_TYPE=vm localintegration
```
Not all tests will work with it (at least for now), but nonetheless we support specifying the RUNTIME_TYPE, so would be good to have this documented.

fgiudici commented on test/helpers.bash:
```diff
@@ -304,9 +303,9 @@ function setup_crio() {
-        --runtimes "$RUNTIME_NAME:$RUNTIME_BINARY:$RUNTIME_ROOT:$RUNTIME_TYPE" \
         -r "$TESTDIR/crio" \
         --runroot "$TESTDIR/crio-run" \
+        --runtimes "$CONTAINER_RUNTIMES" \
```
If we export CONTAINER_RUNTIMES can we also skip the --runtimes parameter here?

haircommander replied:
I found some weirdness without it, but I forget what that was. Let's try!

fgiudici commented on test/helpers.bash:
```diff
@@ -33,15 +33,14 @@ CONTAINER_CONMON_CGROUP=${CONTAINER_CONMON_CGROUP:-pod}
+RUNTIME_BINARY_PATH=$(command -v $CONTAINER_DEFAULT_RUNTIME || true)
+RUNTIME_TYPE=${RUNTIME_TYPE:-oci}
+# export here so direct calls to crio later inherit the variable
+export CONTAINER_RUNTIMES=${CONTAINER_RUNTIMES:-$CONTAINER_DEFAULT_RUNTIME:$RUNTIME_BINARY_PATH:$RUNTIME_ROOT:$RUNTIME_TYPE}
```
Brilliant!

fgiudici reviewed (commented):
LGTM

fidencio commented:
/lgtm

fidencio commented:
/retest

fidencio commented:
/retest

haircommander commented:
conflict resolved can I have my tag back

haircommander commented:
/retest

haircommander commented:
/retest

haircommander commented:
`Kubernetes e2e suite: [sig-network] IngressClass [Feature:Ingress] should set default value on new IngressClass [Serial]`
known flake
/retest

haircommander commented:
ptal @mrunalp @kolyshkin @umohnani8 @giuseppe @saschagrunert

kolyshkin commented on test/restore.bats:
```diff
@@ -100,7 +100,7 @@ function teardown() {
 
 	# simulate reboot with runc state going away
-	"$CONTAINER_RUNTIME" delete -f "$pod_id"
+	"$RUNTIME_BINARY_PATH" --root "$RUNTIME_ROOT" delete -f "$pod_id"
```
so, since you're using `"$RUNTIME_BINARY_PATH" --root "$RUNTIME_ROOT"` from many places, makes sense to have something like

```bash
function runtime() {
   "$RUNTIME_BINARY_PATH" --root "$RUNTIME_ROOT" "$@"
}
```
and use it everywhere. For example, the above would become
```suggestion
	runtime delete -f "$pod_id"
```
which I think helps readability as well as maintenance (making future patches smaller).

kolyshkin commented on test/helpers.bash:
```diff
@@ -35,17 +35,15 @@ CONTAINER_UID_MAPPINGS=${CONTAINER_UID_MAPPINGS:-}
-fi
+# export here so direct calls to crio later inherit the variable
+export CONTAINER_RUNTIMES=${CONTAINER_RUNTIMES:-$CONTAINER_DEFAULT_RUNTIME:$RUNTIME_BINARY_PATH:$RUNTIME_ROOT:$RUNTIME_TYPE}
+CONTAINER_UID_MAPPINGS=${CONTAINER_UID_MAPPINGS:-}
```
Apparently this one is duplicated?

kolyshkin commented on test/helpers.bash:
```diff
@@ -35,17 +35,15 @@ CONTAINER_UID_MAPPINGS=${CONTAINER_UID_MAPPINGS:-}
+export CONTAINER_RUNTIMES=${CONTAINER_RUNTIMES:-$CONTAINER_DEFAULT_RUNTIME:$RUNTIME_BINARY_PATH:$RUNTIME_ROOT:$RUNTIME_TYPE}
+CONTAINER_UID_MAPPINGS=${CONTAINER_UID_MAPPINGS:-}
+CONTAINER_GID_MAPPINGS=${CONTAINER_GID_MAPPINGS:-}
+OVERRIDE_OPTIONS=${OVERRIDE_OPTIONS:-}
```
... as well as these two

kolyshkin commented on test/helpers.bash:
```diff
@@ -35,17 +35,15 @@ CONTAINER_UID_MAPPINGS=${CONTAINER_UID_MAPPINGS:-}
-RUNTIME_BINARY=${RUNTIME_PATH:-$(command -v runc)}
-RUNTIME_ROOT=${RUNTIME_ROOT:-/run/runc}
+RUNTIME_ROOT=${RUNTIME_ROOT:-/run/$CONTAINER_DEFAULT_RUNTIME}
+RUNTIME_BINARY_PATH=$(command -v "$CONTAINER_DEFAULT_RUNTIME" || true)
```
Maybe we should fail here (i.e. remove ` || true` part).

fidencio commented:
@fgiudici, @haircommander, is this still needed?  If so, @haircommander, would you need someone to take it over or do you have the cycles to take a look at @kolyshkin's comments and do a rebase?

haircommander commented:
I also cleaned up a work around fro crun that seems to be out of date (so I don't forget)

haircommander commented:
/retest

kolyshkin commented on test/helpers.bash:
```diff
@@ -198,6 +198,11 @@ function crictl() {
 
+# Run the runtime binary with the specified RUNTIME_ROOT
+function runtime() {
+    "$RUNTIME_BINARY_PATH" --root "$RUNTIME_ROOT" "$@"
```
Hmm, did you mean `"$CONTAINER_RUNTIME"` here? There's no `"$RUNTIME_BINARY_PATH"` (yet?)

kolyshkin replied:
Ah, I see, the order of commits seems to be wrong (or github is showing them in the wrong order), as this ^^^ belongs to the first commit, and `$RUNTIME_BINARY_PATH` is introduced by the third commit.

haircommander replied:
oopsies, updated!

kolyshkin commented:
1. I think that patches 2 and 4 can be combined into one -- here it does not make sense to add `|| true` and then remove it.

2. Patch 2 breaks all the tests, as it removes the variable used by them. This might play bad if someone wants to use e.g. git bisect.

haircommander commented:
>     1. I think that patches 2 and 4 can be combined into one -- here it does not make sense to add `|| true` and then remove it.
> 
>     2. Patch 2 breaks all the tests, as it removes the variable used by them. This might play bad if someone wants to use e.g. git bisect.

okay, I have implemented point 1. For point 2, I have updated the second commit to not change the variable name in the `runtime()` function, and update it in the third commit. A git bisect should allow each commit to pass, now

kolyshkin reviewed (approved):
LGTM, thanks!

haircommander commented:
/test kata-containers

haircommander commented:
@mrunalp @saschagrunert @fidencio @kolyshkin @umohnani8 @QiWang19 plz merge

saschagrunert reviewed (approved):
/lgtm

haircommander commented:
/retest

haircommander commented:
/test kata-containers

haircommander commented:
/override ci/kata-jenkins

haircommander commented:
/test critest_fedora
